File: src/maxheap/operator/insert.cpp
#include <data_structures/maxheap.hpp>

void maxHeap::insert(int value)
{
    Node* node = new Node(value);
    this->value.push_back(node);

    int index = this->value.size() - 1;
        
    if(index == 0) return ;

    do
    {
        int parent = (index - 1) / 2;
        
        if(swapable(parent, index))
        {
            swap(parent, index);
        }

        index = parent;
    }while(index !=0);

}

File: assets/script/minheap/insert.lua
function minHeap.insert(value)
    table.insert(minHeap.heap, value)

    local index = #minHeap.heap
    
    if(index == 0) then 
        return 
    end

    while(index != 0) do 
        local parent = (index - 1) / 2;
        if(minHeap.swapable(parent, index)) then 
            minHeap.swap(parent, index)
        end
        index = parent
    end
end
